standards/HEALTH_CHECK_SPECIFICATION.md

```diff
@@ -0,0 +1,101 @@
+* Changes in check statuses (including the overall app status) must be logged out
+
+### Check statuses
+
+The status of the check can be one of three values:
+
+Status     | Description
+-----------|--------------
+`OK`       | Everything is ok
+`WARNING`  | Things are degraded, but at least partially functioning
+`CRITICAL` | Failure, your app will now be unable to function
+
+Health check endpoint
+---------------------
+
+The health check data is served via the health check endpoint.  This health check data comes in two forms.  The status code provides a quick way to get the overall status of the app, while the JSON body provides further information that might be useful in triaging any issues.
+
+The following requirements must be met by health check implementations:
+
+* All apps must have a health check endpoint
+* The health check endpoint must be available as a `GET /healthcheck` endpoint
+* The health check must return the [appropriate status code](#health-check-status-codes)
+* The health check must return [status information as JSON](#health-check-body)
+
+### Health check status codes
+
+There are three standard status codes that a health check can return:
+
+Status code | Description
+------------|------------------------
+`200`       | I'm fine
+`429`       | I'm warming up or I'm degraded
+`500`       | Kill me
+
+The health check should return 429 until the first checks have been run.  After that, the codes returned should be based on the check status.  It is the responsibility of whatever is monitoring the health check to provide a grace period on startup.
+
+If any check returns a `CRITICAL` status (see [check statuses below](#check-statuses) then the app should transition to a `429`.  After a timeout that can be set per app, the `429` will be downgraded `500`.  This delay in marking the app as `500` is to allow the issue a chance to resolve on its own before moving to a `500` which will result in the app being killed.
+
+If any check returns a `WARNING` status (see [check statuses below](#check-statuses) then the app should transition to a `429`.  In this situation, the timeout does not apply and the app will continue to return a `429` until the status of the checks changes.
+
+Any status code ≥ 400 && != 429 will be treated as a failure.
+
+### Health check body
+
+The JSON body of the health check should implement the following spec:
+
+Field        | Type     | Description
+-------------|----------|---------------
+`status`     | `string` | The overall status of the health check using the values as the [check statuses](#check-statuses)
+`version`    | `string` | The version information of the app (e.g. commit, semver version, go version, build time, etc) as a string.
+`uptime`     | `ms`     | Milliseconds elapsed since the app started <sup>1</sup>
+`start_time` | `ISO8601`<sup>2</sup> | The time the app started in UTC
```

Comment:
For posterity I will just make a note of our conversation here.  The decision was made since people preferred to be able to quickly see Xms to determine if the app had recently been restarted, but would rather have the start time for comparison to things like logs without having to calculate either one in a time pressured situation.